backend/optimizer.py

```diff
@@ -1,46 +1,127 @@
-        return optimize.minimize(self.to_minimize, objective, tol=1.0, method="powell")
-
-    def to_minimize(self, schedule):
+    def optimize(self):
         indices = list(set(chain.from_iterable(
-            map(lambda x: self.producer.schedule[x][1].load_profile.index.values.tolist(), range(0, len(schedule))))))
+            map(lambda x: self.producer.schedule[x][1].load_profile.index.values.tolist(),
+                range(0, len(self.producer.schedule))))))
         indices.sort()
+        self.indices = indices
 
-        consumed = 0
-        produced = 0
-        penalty = 0
-        for t in indices:
+        # Interpolate and differentiate load profiles before optimization.
+        self.differentiated_production = self.differentiate_and_interpolate(self.producer.prediction, indices)
+        for s in self.producer.schedule:
+            self.differentiated_loads.append(self.differentiate_and_interpolate(s[1].load_profile, indices))
+
+        objective = np.zeros(len(self.producer.schedule))
+        # return optimize.minimize(self.to_minimize, objective, tol=1.0, method="trust-ncg")
+        configs = list(map(list, itertools.product([0, 1], repeat=len(self.differentiated_loads))))
+        return self.recursive_binary_optimizer(configs, float('inf'), configs[0])
+
+    def to_minimize(self, schedule: List[float]):
+        produced = []
+        consumed = []
+
+        for t in self.indices:
             consumed_t = 0
-            for i, j in enumerate(schedule):
-                if j > 0.0:
-                    consumed_t += load_profile_utils.power(self.producer.schedule[i][1].load_profile, t)
-            produced_t = load_profile_utils.power(self.producer.prediction, t)
-
-            produced += produced_t
-            consumed += consumed_t
-            if produced_t - consumed_t < 0:
-                penalty += 1
-
-        score = math.fabs(produced - consumed) + penalty * 0
-        return score
+            for i, load in enumerate(self.differentiated_loads):
+                if schedule[i] > 0:
+                    consumed_t += load[t]
+
+            consumed.append(consumed_t)
+            produced.append(self.differentiated_production[t])
+
+        penalty = 0
+        diff = 0
+        for p, c in zip(produced, consumed):
+            diff_t = abs(p - c)
+            diff += diff_t
+            if c > p:
+                penalty += diff_t
+
+        return diff, penalty
+
+    def binary_optimzier(self):
+        configs = map(list, itertools.product([0, 1], repeat=len(self.differentiated_loads)))
+        min_config = None
+        min_config_value = float('inf')
+        for c in configs:
+            diff, penalty = self.to_minimize(c)
+            if diff < min_config_value:
+                min_config_value = diff + penalty
+                min_config = c
+        return min_config
+
+    def recursive_binary_optimizer(self, configs, min_value, min_config):
```

Comment:
this is the (a bit) smarter implementation of the binary optimizer. Whenever we find an input yielding a greater penalty than the currently discovered smallest return value of `to_minimize()`, we know for a fact that no input containing _all_ these input values can be any smaller. Thus, in these cases we remove every configuration containing _all_ inputs of the current one. This might be a bad explanation, but it can be found in the first of the three bottom if-sentences.